### PR TITLE
[luci-interpreter] Add Transpose kernel

### PR DESCRIPTION
**Pull request by struss**
This commit enable `Transpose` kernel on luci-interpreter.

ONE-DCO-1.0-Signed-off-by: KiDeuk Bang <rrstrous@nate.com>

### CODE REVIEW

File: compiler/luci-interpreter/src/kernels/Transpose.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  const int *perm_data = getTensorData<int32_t>(_perm);
+  const int size = _perm->shape().dim(0);
+  params.perm_count = size;
+  for (int i = 0; i < size; i++)
+    params.perm[i] = perm_data[i];
+  switch (_input->element_type())
+  {
+    case DataType::FLOAT32:
+      tflite::reference_ops::Transpose(params, getTensorShape(_input),
+                                       getTensorData<int32_t>(_input), getTensorShape(_output),
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/Transpose.cpp:72]: maybe getTensorData<float> ?

**Comment by jinevening** [comment]
(optional) You can easily add test cases for uint8 using `TYPED_TEST` (see Split.test.cpp).

File: compiler/luci-interpreter/src/kernels/Transpose.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  // Transpose op only supports 1D-5D input arrays.
+  int dims = _input->shape().num_dims();
+  const int *perm_data = getTensorData<int32_t>(_perm);
+
+  assert(_input->shape().num_dims() <= 5);
+  assert(_input->element_type() == _output->element_type());
+
+  assert(_perm->shape().num_dims() == 1);
+  assert(_perm->shape().dim(0) == dims);
+  for (int i = 0; i < dims; i++)
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Transpose.cpp:47]: (optional) You can fuse this loop into the below one.

File: compiler/luci-interpreter/src/kernels/Transpose.test.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      10, 11, 12, 13, 14, 30, 31, 32, 33, 34, 50,  51,  52,  53,  54,  //
+      70, 71, 72, 73, 74, 90, 91, 92, 93, 94, 110, 111, 112, 113, 114, //
+      15, 16, 17, 18, 19, 35, 36, 37, 38, 39, 55,  56,  57,  58,  59,  //
+      75, 76, 77, 78, 79, 95, 96, 97, 98, 99, 115, 116, 117, 118, 119  //
+  };
+  EXPECT_THAT(extractTensorData<float>(output_tensor),
+              ElementsAreArray(ArrayFloatNear(ref_output_data)));
+}
+
+TEST(TransposeTest, LargeFloatCheck)
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Transpose.test.cpp:87]: It is hard to tell from the name how this test is different from the prevoius one. It is better to render the test specifics it its name.

**Comment by s-barannikov** [comment]
> Conflicting files
compiler/luci-interpreter/src/KernelBuilder.cpp
compiler/luci-interpreter/src/KernelBuilder.h
compiler/luci-interpreter/src/kernels/CMakeLists.txt

Sorry for this :)

File: compiler/luci-interpreter/src/kernels/Transpose.test.cpp (hunk 3 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/Transpose.test.cpp:87]: Oh, Sorry. I will fix on some other example.

File: compiler/luci-interpreter/src/kernels/Transpose.cpp (hunk 2 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/Transpose.cpp:47]: Oh! I will fix it!

File: compiler/luci-interpreter/src/kernels/Transpose.cpp (hunk 1 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/Transpose.cpp:72]: I checked that on tensorflow lite kernel source code. but they use int32_t on float and use int8_t on uint8_t. but I think the suggestion that you give looks much better! I will fix it! thanks!
(refer from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/transpose.cc#L117-L132)

**Comment by struss** [comment]
> > Conflicting files
> > compiler/luci-interpreter/src/KernelBuilder.cpp
> > compiler/luci-interpreter/src/KernelBuilder.h
> > compiler/luci-interpreter/src/kernels/CMakeLists.txt
> 
> Sorry for this :)

Never mind. I will resolve the conflict and make new commit!

File: compiler/luci-interpreter/src/kernels/Transpose.cpp (hunk 1 above)
Comment (binarman) [reply compiler/luci-interpreter/src/kernels/Transpose.cpp:72]: This is the way tflite doing it:
```
  switch (op_context.input->type) {
    case kTfLiteFloat32:
    case kTfLiteInt32:
      if (kernel_type == kGenericOptimized) {
        TF_LITE_TRANSPOSE(optimized_ops, int32_t);
      } else {
        TF_LITE_TRANSPOSE(reference_ops, int32_t);
      }
      break;
```
Though, I am not sure this is fully c++ standard safe =/

Comment (binarman) [reply compiler/luci-interpreter/src/kernels/Transpose.cpp:72]: Comment from reference_ops.h
```
  // Transpose kernel only does rearranging values not numeric evaluations on
  // each cell. It's safe to implement per size of scalar type and this trick
  // keeps the total code size in a reasonable range.
```

File: compiler/luci-interpreter/src/kernels/Transpose.test.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+template <typename T> class TransposeTest : public ::testing::Test
+{
+};
+
+using DataTypes = ::testing::Types<float, uint8_t>;
+TYPED_TEST_CASE(TransposeTest, DataTypes);
+
+TYPED_TEST(TransposeTest, Small3D)
+{
+  Check<TypeParam>({2, 3, 4}, {3}, {4, 2, 3}, {0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11,
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Transpose.test.cpp:58]: Could you please add inline hints? Like `/*input_shape=*/`?
The code would be more clear.

Comment (struss) [reply compiler/luci-interpreter/src/kernels/Transpose.test.cpp:58]: Oh, Okay I will add a hint on all test values.

**Comment by s-barannikov** [comment]
Please revert change of `tests/nnapi/specs/Ex/shape_ex_dynamic_nnfw.mod.py`

**Comment by struss** [comment]
> Please revert change of `tests/nnapi/specs/Ex/shape_ex_dynamic_nnfw.mod.py`

Sorry, just follow what format checker said. I reverted it.

**Comment by s-barannikov** [comment]
@struss FYI https://github.com/Samsung/ONE/pull/1890

**Comment by struss** [comment]
> @struss FYI #1890

Oh, Thanks! I will check that PR.

**Comment by s-barannikov** [comment]
It's already merged. If format checker fails again, please rebase onto master.

**Comment by seanshpark** [comment]
@struss , please rebase to latest master

**Comment by struss** [comment]
> @struss , please rebase to latest master

Okay! Done it! Thank you!

**Pull request merged**